api/include/opentelemetry/trace/global_provider.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/trace/noop.h"
+#include "opentelemetry/trace/tracer_provider.h"
+
+namespace opentelemetry
+{
+namespace trace
+{
+
+class DefaultTracerProvider final : public opentelemetry::trace::TracerProvider
+{
+public:
+  opentelemetry::trace::NoopTracer *const GetTracer(
+      nostd::string_view library_name,
+      nostd::string_view library_version  = "",
+      const TracerOptions &tracer_options = {}) override
+  {
+    return tracer_.get();
+  }
+
+private:
+  DefaultTracerProvider()
+      : tracer_{nostd::unique_ptr<opentelemetry::trace::NoopTracer>(
+            new opentelemetry::trace::NoopTracer)}
+  {}
+  nostd::unique_ptr<opentelemetry::trace::NoopTracer> tracer_;
+
+  friend class Provider;
+};
+
+class Provider
+{
+public:
+  static opentelemetry::trace::TracerProvider *GetTracerProvider()
+  {
+    return Ptr()->load(std::memory_order_acquire);
+  }
+
+  static void SetTracerProvider(TracerProvider *tp)
+  {
+    if (auto old = Ptr()->exchange(tp, std::memory_order_release))
+    {
+      delete old;
```

Comment:
Exchange fixes part of the race, but there's still a race between Get and Set:

* Get a provider.
* Another thread calls Set and deletes it.
* Try to use the now deleted provider.

I think our options here are:
* Make TracerProvider refcounted to solve the lifetime issue (e.g. GetTracerProvider returns a shared_ptr)

or

* We get out of the memory lifetime business: don't delete anything. If a developer wants to set the global provider, it's up to them to manage its lifetime. The DefaultTracerProvider is a singleton and we never delete it.

Reply:
Yeah, good points.  I think I'm for letting the user deal with the lifetime for now, and record the awkward ergonomics of the API in an issue that could be addressed later.  I definitely see some value in having this be managed, but it can be implemented as a follow up.

According to the spec these kinds of configurations should happen only once per process, so the chances of encouraging a leak here are low. Maybe we should enforce this programmatically with a `static bool provider_was_set_` member?

Another idea is we return the pointer we are replacing to the old TracerProvider in the call to encourage users to clean up after themselves like:
```
  ...
  auto provider = new TracerProvider(...);
  delete opentelemetry::trace::Provider::SetTracerProvider(provider);
  ...
```
  
This might encourage folks to run into the exact scenario we are trying to address here, which isn't the best ergonomically.

Reply:
I agree.

I think setting the global trace provider once during start-up should be the common case.

Reply:
@rnburn Further to the meeting today, what is actually the global state we want to be able to set here? TracerProvider? Factory? Tracer?

Reply:
The global providers are meant to replace the global tracer/meters.

> These factories (TracerProvider and MeterProvider) replace the global Tracer / Meter singleton objects as ubiquitous points to request Tracer and Meter instances.

https://github.com/open-telemetry/oteps/blob/master/text/0016-named-tracers.md#explanation

Reply:
Thanks for the additional context. Sounds like the API just needs the global TraceProvider.

Can we mark this conversation resolved?

Reply:
I believe there's an underlying SDK object that needs to be swapped out under the hood here.

The main functionality is: when you set your provider, there should be a stop the world moment where you iterate through all your default (i.e. noop) tracers/meters and swap their internals out with an actual implementation.  In this case, we will want to generate a new tracer object like we do in the SDK, but have it point to a noop SDK implementation. This means we can pull the logic in from the SDK provider to the API provider.

Thoughts?

Reply:
I'm going to punt on the proxy idea for now in lieu of getting this closed up. I'll get a PR in that implements the proxy in as a follow on.